### PR TITLE
Document our security model

### PR DESCRIPTION
**Pull request by brian-brazil**

### CODE REVIEW

**Comment by RichiH** [comment]
`secret` and `string` both start with s and are the same length. `confidential` or some such might be less error-prone.

File: content/docs/operating/security.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+---
+title: Security
+sort_rank: 4
+---
+
+# Security Model
+
+Prometheus is sophisticated system with many components and many integrations with
```
Comment (RichiH) [inline content/docs/operating/security.md]: a system

File: content/docs/operating/security.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+other systems. It can be deployed in a varietry of trusted and untrusted
+enviroments.
```
Comment (RichiH) [inline content/docs/operating/security.md]: n

File: content/docs/operating/security.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+This page describes the general security assumptions of Prometheus and the
+attack vectors that some configurations may enable.
+
+As with any complex systems it is not possible to gaurantee that there are no
+bugs. If you find a security bug, please file it in the issue tracker of the
+relevant component.
```
Comment (RichiH) [inline content/docs/operating/security.md]: Something about responsible disclosure in case of bad bugs?

File: content/docs/operating/security.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+### Prometheus
+
+It is presumed that untrusted users have access to the Prometheus HTTP endpoint
+and logs. They have access to all time series information contained in the
+database, plus a variety of operational/debugging information.
+
+It is presumed that only trusted users have the ability to change the command
```
Comment (RichiH) [inline content/docs/operating/security.md]: It is _also_ presumed.

File: content/docs/operating/security.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+possible for a target to expose data that impersonates a different target.  The
+`honor_labels` option removes this protection, as can certain relabelling
+setups.
+
+At present the HTTP API grants the ability to reload configuration files,
+shutdown the server, and delete time seres. This will be [changed in Prometheus
+2.0](https://github.com/prometheus/prometheus/issues/2173).
+
+The remote read feature allows anyone with HTTP access to send queries to the
+remote read endpoint. If for example the PromQL queries were ending directly
```
Comment (RichiH) [inline content/docs/operating/security.md]: That sentence broken.

File: content/docs/operating/security.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+As with any complex systems it is not possible to gaurantee that there are no
+bugs. If you find a security bug, please file it in the issue tracker of the
+relevant component.
+
+### Prometheus
+
+It is presumed that untrusted users have access to the Prometheus HTTP endpoint
+and logs. They have access to all time series information contained in the
+database, plus a variety of operational/debugging information.
```
Comment (RichiH) [inline content/docs/operating/security.md]: and operations, at least in 1.x

**Comment by brian-brazil** [comment]
"Secret" is the standard security term for this. Confidential has a different implication (for example my bank statements are confidential, but they are not a security secret).

File: content/docs/operating/security.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+Prometheus and its components do not provide any server-side
+authentication, authorisation or encryption. If you require this, it is
+recommended to use a reverse proxy.
+
+Various Prometheus components support client-side authentication and
+encryption. If TLS client support is offered, there is often also an option
+called `insecure_skip_verify` which skips SSL verification.
+
+## Secrets
```
Comment (RichiH) [inline content/docs/operating/security.md]: Maybe rename to confidential or such, as per comment in PR.

File: content/docs/operating/security.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+components to expose their configuration over their HTTP endpoint.
+
+Secrets from other sources used by dependencies (e.g. the `AWS_SECRET_KEY`
+enviroment vabiable as used by EC2 service discovery) may end up exposed due to
+code outside of our control or due to functionality that happens to expose
+wherever it is stored.
+
+## Denial of Service
+
+While there are some mitigations in place for excess load, if too many
```
Comment (RichiH) [inline content/docs/operating/security.md]: Or just one really large query.

File: content/docs/operating/security.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+automatically restart on failure.
+
+## Libraries
+
+This document considers vanilla binaries built from the stock source code.
+Information presented here does not apply if you modify Prometheus source code,
+or use Prometheus internals (beyond the official client library APIs) in your
+own code.
+
+## Build Process
```
Comment (RichiH) [inline content/docs/operating/security.md]: and resulting binaries.

File: content/docs/operating/security.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+The build pipeline for Prometheus runs on 3rd party providers to which many
+members of the Prometheus development team has access. If you are concerned
```
Comment (RichiH) [inline content/docs/operating/security.md]: access, but the real control lies with the 3rd parties which are totally independent. This should be worded more strongly.

**Comment by RichiH** [comment]
I am aware of that, but on the other hand, that is seldomly mixed with `string`.

File: content/docs/operating/security.md (hunk 6 above)
Comment (brian-brazil) [inline content/docs/operating/security.md]: Covered a few paragraphs down.

Comment (RichiH) [inline content/docs/operating/security.md]: I know, but it's 1-2 more words and as it's written right now, it's not 100% correct.

File: content/docs/operating/security.md (hunk 3 above)
Comment (brian-brazil) [inline content/docs/operating/security.md]: We've had by my count 2 security bugs (one XSS, one secret exposure), I'd currently prefer to risk a bad bug being public early than pointing more support load directly at -team.

**Comment by brian-brazil** [comment]
Done, bar the ones I commented on.

**Comment by RichiH** [comment]
I would like to get a third voice on `string` and `secret`; plus, adding "operations" is cheap and makes things more exact.

Else, LGTM.

File: content/docs/operating/security.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+or use Prometheus internals (beyond the official client library APIs) in your
+own code.
+
+## Build Process
+
+The build pipeline for Prometheus runs on 3rd party providers to which many
+members of the Prometheus development team and the staff of those providers
+have access. If you are concerned about the exact provenance of your binaries,
+it is recommended to build them yourself rather than relying on the
+pre-binaries provided by the project.
```
Comment (Gouthamve) [inline content/docs/operating/security.md]: pre-built

Comment (brian-brazil) [inline content/docs/operating/security.md]: Fixed

File: content/docs/operating/security.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
+---
+title: Security
+sort_rank: 4
+---
+
+# Security Model
+
+Prometheus is a sophisticated system with many components and many integrations
+with other systems. It can be deployed in a varietry of trusted and untrusted
```
Comment (juliusv) [inline content/docs/operating/security.md]: varietry -> variety

File: content/docs/operating/security.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+relabelling may grant some of this control to anyone who can modify data in
+that service discovery system.
+
+Scraped targets may be run by untrusted users. It should not by default be
+possible for a target to expose data that impersonates a different target.  The
+`honor_labels` option removes this protection, as can certain relabelling
+setups.
+
+At present the HTTP API grants the ability to reload configuration files,
+shutdown the server, and delete time seres. This will be [changed in Prometheus
```
Comment (juliusv) [inline content/docs/operating/security.md]: seres -> series

File: content/docs/operating/security.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+2.0](https://github.com/prometheus/prometheus/issues/2173).
+
+The remote read feature allows anyone with HTTP access to send queries to the
+remote read endpoint. If for example the PromQL queries were ending up directly
+run against relational database, then anyone with tha ability to send queries
```
Comment (juliusv) [inline content/docs/operating/security.md]: against -> against a

tha -> the

File: content/docs/operating/security.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+## Alertmanager
+
+Any user with access to the Alertmanager HTTP endpoint has access to its data.
+They can create and resolve alerts. They can create, modify and delete
+silences.
+
+Where notifications are sent to is determined by the configuration file. With
+certain templating setups it is possible for notifications to end up at an
+alert-defined destination. For example if notifications use an alert label as
+the destination email address, anyone who can send alerts to the alertmanager
```
Comment (juliusv) [inline content/docs/operating/security.md]: alertmanager -> Alertmanager

File: content/docs/operating/security.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+exporters can make them send requests to arbitrary endpoints. As they also
+support client-side authentication, this could lead to a leak of secrets such
+as HTTP Basic Auth passwords or SNMP community strings. Challenge-response
+authentication mechanisms such as TLS are not affected by this.
+
+## Client Libraries
+
+Client libaries are intended to be included in users' applications.
+
+If using a client libary provided HTTP handler, it should not be possible for
```
Comment (juliusv) [inline content/docs/operating/security.md]: libary -> library

client library provided -> client-library-provided

File: content/docs/operating/security.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+components to expose their configuration over their HTTP endpoint.
+
+Secrets from other sources used by dependencies (e.g. the `AWS_SECRET_KEY`
+enviroment vabiable as used by EC2 service discovery) may end up exposed due to
+code outside of our control or due to functionality that happens to expose
+wherever it is stored.
+
+## Denial of Service
+
+While there are some mitigations in place for excess load or expensive queries.
```
Comment (juliusv) [inline content/docs/operating/security.md]: A sentence starting with "While ..." (subordinate clause) needs a main clause following it (after a comma).

Just omit the "While" and it's correct :)

File: content/docs/operating/security.md [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+However if too many or too expensive queries/metrics are provided components
```
Comment (juliusv) [inline content/docs/operating/security.md]: Commas after "However" and "provided" would make this sentence easier to read.

File: content/docs/operating/security.md [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+will fall over. It is more likely that a component will be accidentally taken
+out by a trusted user than malicious action.
+
+It is the responsibility of the the user to ensure they provide components with
+sufficient resources including CPU, RAM, disk space, iops, file descriptors,
```
Comment (juliusv) [inline content/docs/operating/security.md]: iops -> IOPS

File: content/docs/operating/security.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+## Libraries
+
+This document considers vanilla binaries built from the stock source code.
+Information presented here does not apply if you modify Prometheus source code,
+or use Prometheus internals (beyond the official client library APIs) in your
+own code.
+
+## Build Process
+
+The build pipeline for Prometheus runs on 3rd party providers to which many
```
Comment (juliusv) [inline content/docs/operating/security.md]: 3rd part -> third-party

**Comment by juliusv** [comment]
I prefer `secret` over `confidential`.

**Comment by brian-brazil** [comment]
Updated.

**Comment by RichiH** [comment]
I fixed the final few typos in 8cad96e04b3a9f01b8422f2c0fc1bf072b17c50d to make this quicker.

**Pull request merged**